src/FullMarginLib.sol

```diff
@@ -51,7 +55,10 @@ library FullMarginLib {
 
         uint80 newAmount = account.collateralAmount - amount;
         account.collateralAmount = newAmount;
-        if (newAmount == 0) {
+
+        // only reset the collateralId if the collateral amount is 0 and there is no short position
+        // if tokenId is non-empty, we keep collatId here so mismatched collateral cannot be added
+        if (newAmount == 0 && account.tokenId == 0) {
```

Comment:
This is the bug fix to "minting with wrong collateral issue" found both by Certora rule (also found in Delvir0's audit)

Reply:
This does indeed fixes the wrong collateral issue. 
Do note that at the same time it opens the same problem as "[L-04] Assumption on resetting collateral will fail in cases where tokens do not support 0 transfers".
When a user calls remove collateral while still holding an option, the user might not be able to reset the collateralId anymore afterwards (since we would have to call removeCollateral(0) and some stables do not support 0 transfers).
Although the likelihood of this happening is higher than [L-04] the impact is still low where it would be a user inconvenient.

Reply:
thanks for the comment. I'd consider just add that line in [L-04] and reset the collateral ID if ending amount is 0, seems like it introduces quite some edge cases.